File: Max min Height.py
class Solution():
    def isValid(self, arr, val, w, m):
        n = len(arr)
        difSum = [0]*n
        pref = 0
        for i in range(n):
            pref -= difSum[i]
            req = max(0, val - arr[i]-pref)
            m-=req
            pref += req
            if i+w< n:
                difSum[i+w] +=  req
        if m < 0:
            return False
        return True

    def maximizeMinHeight(self, n, k, w,a):
        l = 0
        r = 10**9+1
        while l < r-1:
            mid = (l+r)//2
            if self.isValid(a, mid, w, k):
                l = mid
            else:
                r = mid - 1
        while r>=l:
            if self.isValid(a, r, w, k):
                return r
            r -= 1

File: Max min Height.cpp
class Solution
{
    public:
        bool isvalid(vector <int> &a,int n,int k,int w,long long int mid)
        {
            long long int ps[n];
            fill(ps,ps+n,0ll);
            long long int ans=0;
            for(long long int i=0;i<n;i++)
            {
                if(i-1>=0)
                    ps[i]+=ps[i-1];
                if(a[i]+ps[i]<mid)
                {
                    long long int e=mid-a[i]-ps[i];
                    ans+=e;
                    ps[i]+=e;
                    if(i+w<n)
                        ps[i+w]-=e;
                }
            }
            return (ans<=k);
            // for()
        }
        long long int maximizeMinHeight(int n,int k,int w,vector <int> &a)
        {
            int ans=0;
            long long int l=0,h=1e14;
            while(l<=h)
            {
                long long int mid=(l+h)/2;
                if(isvalid(a,n,k,w,mid))
                {
                    l=mid+1;
                    ans=mid;
                }
                else
                {
                    h=mid-1;
                }
            }
            return ans;
        }
};
